openzl-crypto/src/poseidon/mod.rs

```diff
@@ -0,0 +1,634 @@
+        S::permute_without_first_round(&self.additive_round_keys, &self.mds_matrix, state, compiler)
+    }
+
+    /// Computes the first round borrowing the `input` and `domain_tag` returning the [`State`]
+    /// after the first round. This method does not check that `N + 1 = S::WIDTH`.
+    #[inline]
+    fn first_round_with_domain_tag_unchecked<const N: usize>(
+        &self,
+        domain_tag: &S::Field,
+        input: [&S::Field; N],
+        compiler: &mut COM,
+    ) -> State<S, COM> {
+        let keys = self.additive_keys(0);
+        S::first_round_with_domain_tag_unchecked(
+            domain_tag,
+            keys,
+            &self.mds_matrix,
+            input,
+            compiler,
+        )
+    }
+}
+
+impl<S, COM> Constant<COM> for Permutation<S, COM>
+where
+    S: Specification<COM> + Constant<COM>,
+    S::Type: Specification<ParameterField = Const<S::ParameterField, COM>>,
+    S::ParameterField: Constant<COM>,
+{
+    type Type = Permutation<S::Type>;
+
+    #[inline]
+    fn new_constant(this: &Self::Type, compiler: &mut COM) -> Self {
+        Self::new_unchecked(
+            this.additive_round_keys
+                .iter()
+                .map(|e| e.as_constant(compiler))
+                .collect(),
+            this.mds_matrix
+                .iter()
+                .map(|e| e.as_constant(compiler))
+                .collect(),
+        )
+    }
+}
+
+impl<S, COM> Decode for Permutation<S, COM>
+where
+    S: Specification<COM>,
+    S::ParameterField: Decode,
+{
+    type Error = <S::ParameterField as Decode>::Error;
+
+    #[inline]
+    fn decode<R>(mut reader: R) -> Result<Self, DecodeError<R::Error, Self::Error>>
+    where
+        R: Read,
+    {
+        Ok(Self::new_unchecked(
+            (0..S::ADDITIVE_ROUND_KEYS_COUNT)
+                .map(|_| Decode::decode(&mut reader))
+                .collect::<Result<_, _>>()?,
+            (0..S::MDS_MATRIX_SIZE)
+                .map(|_| Decode::decode(&mut reader))
+                .collect::<Result<_, _>>()?,
+        ))
+    }
+}
+
+impl<S, COM> Encode for Permutation<S, COM>
+where
+    S: Specification<COM>,
+    S::ParameterField: Encode,
+{
+    #[inline]
+    fn encode<W>(&self, mut writer: W) -> Result<(), W::Error>
+    where
+        W: Write,
+    {
+        for key in self.additive_round_keys.iter() {
+            key.encode(&mut writer)?;
+        }
+        for entry in self.mds_matrix.iter() {
+            entry.encode(&mut writer)?;
+        }
+        Ok(())
+    }
+}
+
+impl<S, COM> PseudorandomPermutation<COM> for Permutation<S, COM>
+where
+    S: Specification<COM>,
+{
+    type Domain = State<S, COM>;
+
+    #[inline]
+    fn permute(&self, state: &mut Self::Domain, compiler: &mut COM) {
+        self.full_round(0, state, compiler);
+        self.permute_without_first_round(state, compiler);
+    }
+}
+
+impl<S, COM> Sample for Permutation<S, COM>
```

Comment:
This Sample is deterministic, so maybe we should impl `Default` instead. @bhgomes thoughts?

Reply:
The reason we are using `Sample` is because we have a chain of sample algorithms and so it's expecting a `Sample` implementation. This can be rewritten in future versions.